foundation/auth/src/project.rs

```diff
@@ -93,11 +94,22 @@ pub async fn create_token_source_from_project(
 ) -> Result<Box<dyn TokenSource>, error::Error> {
     match project {
         Project::FromFile(file) => create_token_source_from_credentials(file, &config).await,
-        Project::FromMetadataServer(_) => {
-            let ts = ComputeTokenSource::new(&config.scopes_to_string(","))?;
-            let token = ts.token().await?;
-            Ok(Box::new(ReuseTokenSource::new(Box::new(ts), token)))
-        }
+        Project::FromMetadataServer(_) => create_token_source_from_metadata_server(&config).await,
+    }
+}
+
+/// Create a token source that use the metadata server to generate tokens.
+pub async fn create_token_source_from_metadata_server(
+    config: &Config<'_>,
+) -> Result<Box<dyn TokenSource>, error::Error> {
+    if let Some(audience) = config.audience {
```

Comment:
The current implementation always uses ComputeIdentitySource when used via [google-cloud-default](https://github.com/golang/oauth2/blob/2323c81c8dba82e8650ed3a24a1a5667e293af38/google/default.go#L164).
If you look at the [Go library](https://github.com/golang/oauth2/blob/master/google/default.go#L160) , `ComputeTokenSource` is used in the GCP environment, so we would like to have a special option to specify `ComputeIdentitySource`. For example, by adding a flag to the Config as follows.
```rust
 if let Some(audience) = config.audience && config.use_identity_token_source {
 // use ComputeIdentitySource
} else {
  // use ComputeTokenSource
}

```

Reply:
I hadn't realize it was modeled after the go library. In that case, I think it's better that the logic for "automatic" source selection stays close to the original implementation.

I propose we don't change `create_token_source_from_project` and instead we provide a top module level constructor function that one can use explicitly. In the same spirit as this:

https://github.com/golang/oauth2/blob/2323c81c8dba82e8650ed3a24a1a5667e293af38/google/jwt.go#L27

What do you think ?

Reply:
Actually the idtoken library, which purpose is to call other services, [uses the `identity` endpoint](https://github.com/googleapis/google-api-go-client/blob/f06cac5563b71f297a8ee726382fadc106995924/idtoken/compute.go#L42). Maybe we should add an `idtoken` module that implement the same [selection token source logic from the credientials](https://github.com/googleapis/google-api-go-client/blob/f06cac5563b71f297a8ee726382fadc106995924/idtoken/idtoken.go#L101) ?